js/whitebit.js

```diff
@@ -0,0 +1,425 @@
+        const timestamp = this.parse8601 (this.safeString (result, 'lastUpdateTimestamp'));
+        const orderbook = this.parseOrderBook (result, timestamp);
+        orderbook['nonce'] = this.safeInteger (result, timestamp);
+        return orderbook;
+    }
+
+    async fetchTrades (symbol, since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const request = {
+            'market': market['id'],
+            'lastId': 1, // todo add since
+        };
+        if (limit !== undefined) {
+            request['limit'] = limit; // default = 50, maximum = 10000
+        }
+        const response = await this.publicV1GetHistory (this.extend (request, params));
+        const result = this.safeValue (response, 'result');
+        return this.parseTrades (result, market, since, limit);
+    }
+
+    parseTrade (trade, market = undefined) {
+        const timestamp = this.safeTimestamp (trade, 'time');
+        const price = this.safeFloat (trade, 'price');
+        const amount = this.safeFloat (trade, 'amount');
+        const id = this.safeString (trade, 'id');
+        const side = this.safeString (trade, 'type');
+        const symbol = market['symbol'];
+        let cost = undefined;
+        if (amount !== undefined && price !== undefined) {
+            cost = amount * price;
+        }
+        return {
+            'info': trade,
+            'timestamp': timestamp,
+            'datetime': this.iso8601 (timestamp),
+            'symbol': symbol,
+            'id': id,
+            'order': undefined,
+            'type': undefined,
+            'takerOrMaker': undefined,
+            'side': side,
+            'price': price,
+            'amount': amount,
+            'cost': cost,
+            'fee': undefined,
+        };
+    }
+
+    async fetchOHLCV (symbol, timeframe = '1m', since = undefined, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const request = {
+            'market': market['id'],
+            'interval': this.timeframes[timeframe],
+        };
+        if (since !== undefined) {
+            request['start'] = parseInt(since / 1000);
+        }
+        if (limit !== undefined) {
+            request['limit'] = limit; // default == max == 500
+        }
+        const response = await this.publicV1GetKline (this.extend (request, params));
+        const result = this.safeValue (response, 'result');
+        return this.parseOHLCVs (result, market, timeframe, since, limit);
+    }
+
+    parseOHLCV (ohlcv, market = undefined, timeframe = '1m', since = undefined, limit = undefined) {
+        return [
+            ohlcv[0] * 1000,
+            parseFloat (ohlcv[1]),
+            parseFloat (ohlcv[2]),
+            parseFloat (ohlcv[3]),
+            parseFloat (ohlcv[4]),
+            parseFloat (ohlcv[5]),
+        ];
```

Comment:
Проблема ещё в том, что порядок здесь не OHLCV, т.к. получается, что open > high, чего не может быть по определению. К сожалению, этот endpoint не документирован...

Reply:
```
whitebit.fetchOHLCV (ETH/BTC)
fetch:
 whitebit GET https://whitebit.com/api/v1/public/kline?market=ETH_BTC&interval=1m
Request:
 {}
 undefined

handleRestResponse:
 whitebit GET https://whitebit.com/api/v1/public/kline?market=ETH_BTC&interval=1m 200 OK
Response:
 {
  'Cache-Control': 'no-cache, private',
  'Cf-Cache-Status': 'DYNAMIC',
  'Cf-Ray': '525670acaf13dfb7-FRA',
  Connection: 'close',
  'Content-Encoding': 'gzip',
  'Content-Type': 'application/json',
  Date: 'Mon, 14 Oct 2019 03:24:24 GMT',
  'Expect-Ct': 'max-age=604800, ' +
    'report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"',
  Server: 'cloudflare',
  'Set-Cookie': '__cfduid=dedbf3e785296c6421001fac3482d1b9d1571023464; expires=Tue, ' +
    '13-Oct-20 03:24:24 GMT; path=/; domain=.whitebit.com; HttpOnly; ' +
    'Secure',
  'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload',
  'Transfer-Encoding': 'chunked',
  'X-Content-Type-Options': 'nosniff',
  'X-Frame-Options': 'SAMEORIGIN',
  'X-Xss-Protection': '1; mode=block'
}
 {"success":true,"message":"","result":[[1571011200,"0.021874","0.02187","0.021878","0.02187","0.229","0.005009109"],[1571011260,"0.021871","0.021856","0.021877","0.021851","1.001","0.021880405"],[1571011320,"0.021863","0.021861","0.021872","0.021861","1.508","0.032970594"],[1571011380,"0.021863","0.021879","0.021879","0.021862","0.729","0.015942249"],[1571011440,"0.021868","0.02188","0.02188","0.021865","0.78","0.017058762"],[1571011500,"0.021881","0.021888","0.021897","0.021881","1.937","0.042400161"],[1571011560,"0.021885","0.021884","0.02189","0.021881","2.926","0.064037384"],[1571011620,"0.021888","0.021901","0.02191","0.021888","4.328","0.094770108"],[1571011680,"0.021908","0.021897","0.021911","0.021895","2.788","0.061075977"],[1571011740,"0.021887","0.021891","0.021891","0.021887","0.515","0.011273139"],[1571011800,"0.021894","0.021884","0.021896","0.021884","2.321","0.050809214"],[1571011860,"0.021887","0.021893","0.021893","0.021878","2.567","0.056182671"],[1571011920,"0.021887","0.021882","0.021895","0.021881","2.537","0.055529751"],[1571011980,"0.021883","0.02188","0.021889","0.021879","1.125","0.024618205"],[1571012040,"0.021885","0.021885","0.021889","0.021885","2.068","0.045263294"],[1571012100,"0.021888","0.021892","0.021892","0.021884","1.861","0.040737343"],[1571012160,"0.02189","0.021895","0.021901","0.02189","3.329","0.072894377"],[1571012220,"0.021894","0.021888","0.021894","0.021882","1.699","0.037191366"],[1571012280,"0.021885","0.021892","0.021893","0.021885","0.977","0.021385069"],[1571012340,"0.021886","0.021893","0.021893","0.021883","1.222","0.026746747"],[1571012400,"0.021891","0.021884","0.021891","0.021883","0.868","0.018997371"],[1571012460,"0.021885","0.021893","0.021893","0.021885","1.612","0.035283716"],[1571012520,"0.021889","0.021894","0.021896","0.021889","1.082","0.023687838"],[1571012580,"0.021894","0.021904","0.021908","0.021894","1.534","0.033601672"],[1571012640,"0.021904","0.021903","0.021908","0.021899","2.558","0.056029162"],[1571012700,"0.021906","0.021898","0.021907","0.021888","2.963","0.064887763"],[1571012760,"0.021901","0.021897","0.021909","0.021897","4.447","0.097411834"],[1571012820,"0.021899","0.021899","0.021899","0.021893","3.371","0.073813581"],[1571012880,"0.021895","0.021887","0.021896","0.021884","1.631","0.035704851"],[1571012940,"0.021885","0.021877","0.021886","0.021876","0.909","0.019888844"],[1571013000,"0.021882","0.021899","0.0219","0.021882","0.683","0.014948238"],[1571013060,"0.021893","0.021881","0.021896","0.021875","2.229","0.048787789"],[1571013120,"0.021878","0.021884","0.021887","0.021875","0.925","0.020239126"],[1571013180,"0.021887","0.0219","0.0219","0.021887","1.077","0.023581819"],[1571013240,"0.0219","0.02189","0.021903","0.02189","2.574","0.056360615"],[1571013300,"0.021892","0.021899","0.021899","0.021892","2.254","0.049354724"],[1571013360,"0.021901","0.021906","0.021907","0.021899","4.748","0.104000057"],[1571013420,"0.021906","0.021906","0.021908","0.0219","6.003","0.131484634"],[1571013480,"0.021901","0.021899","0.021902","0.021899","4.264","0.093383168"],[1571013540,"0.021899","0.021897","0.021905","0.021897","3.355","0.073478632"],[1571013600,"0.021898","0.021886","0.021903","0.021886","2.368","0.051856448"],[1571013660,"0.021887","0.021891","0.021891","0.02188","0.805","0.017616225"],[1571013720,"0.021896","0.0219","0.021906","0.021895","2.912","0.063770558"],[1571013780,"0.0219","0.021896","0.0219","0.021895","1.089","0.023847019"],[1571013840,"0.0219","0.0219","0.021906","0.0219","1.103","0.024157818"],[1571013900,"0.021902","0.0219","0.021908","0.0219","0.617","0.013514036"],[1571013960,"0.0219","0.021898","0.021906","0.021898","3.591","0.078651302"],[1571014020,"0.0219","0.021905","0.021909","0.021895","0.782","0.017129986"],[1571014080,"0.021906","0.021902","0.021906","0.0219","1.547","0.033884146"],[1571014140,"0.021902","0.021905","0.02191","0.021898","0.988","0.02164114"],[1571014200,"0.021904","0.021908","0.021911","0.021902","2.256","0.049422775"],[1571014260,"0.021907","0.021904","0.021907","0.021901","0.694","0.015202605"],[1571014320,"0.021908","0.02191","0.02191","0.021905","0.552","0.012092446"],[1571014380,"0.021907","0.021912","0.021912","0.021907","0.533","0.01167758"],[1571014440,"0.021916","0.021911","0.021923","0.021911","1.721","0.037719861"],[1571014500,"0.021912","0.021911","0.021915","0.021907","0.681","0.014921955"],[1571014560,"0.021909","0.021915","0.021916","0.021905","0.65","0.014241764"],[1571014620,"0.021915","0.021916","0.021919","0.021914","0.85","0.018628004"],[1571014680,"0.021915","0.021911","0.021915","0.021911","0.575","0.012600008"],[1571014740,"0.021914","0.021913","0.021915","0.021912","1.019","0.022330259"],[1571014800,"0.021911","0.021896","0.021911","0.021894","1.572","0.034432344"],[1571014860,"0.021893","0.021904","0.021905","0.021885","2.656","0.058155204"],[1571014920,"0.021905","0.021914","0.021916","0.021905","0.636","0.013934473"],[1571014980,"0.021916","0.021909","0.021916","0.021909","0.375","0.00821656"],[1571015040,"0.021908","0.021905","0.021908","0.021903","1.688","0.03697669"],[1571015100,"0.021906","0.021907","0.021907","0.021904","1.465","0.032091094"],[1571015160,"0.021907","0.021907","0.021907","0.021901","0.428","0.009375507"],[1571015220,"0.021912","0.021923","0.021926","0.021912","1.186","0.025997295"],[1571015280,"0.021923","0.021922","0.021924","0.021917","0.583","0.012780018"],[1571015340,"0.021921","0.021922","0.021926","0.021921","1.999","0.043825255"],[1571015400,"0.021924","0.021924","0.021927","0.021923","2.249","0.049308312"],[1571015460,"0.02192","0.021914","0.021925","0.021906","1.189","0.02606072"],[1571015520,"0.021908","0.021908","0.021911","0.021908","3.229","0.070744478"],[1571015580,"0.021911","0.02191","0.021911","0.021901","2.092","0.04582629"],[1571015640,"0.021916","0.021913","0.02192","0.021912","1.562","0.034231361"],[1571015700,"0.021912","0.021915","0.021916","0.021912","3.131","0.068611639"],[1571015760,"0.021914","0.021908","0.021917","0.021908","2.195","0.048099245"],[1571015820,"0.021906","0.021917","0.021918","0.021905","0.817","0.01790055"],[1571015880,"0.021917","0.021915","0.021921","0.021915","0.757","0.01659139"],[1571015940,"0.021915","0.021921","0.021925","0.021915","1.016","0.022269981"],[1571016000,"0.021922","0.021919","0.021926","0.021919","0.616","0.013503659"],[1571016060,"0.021918","0.021918","0.021918","0.021916","0.513","0.011243815"],[1571016120,"0.021919","0.021915","0.021924","0.021914","0.583","0.012779344"],[1571016180,"0.021914","0.021914","0.021919","0.021909","0.916","0.020072564"],[1571016240,"0.021916","0.02191","0.021916","0.02191","2.765","0.060589302"],[1571016300,"0.021912","0.021911","0.021914","0.021908","3.12","0.068362574"],[1571016360,"0.021911","0.021921","0.021921","0.02191","1.447","0.031707486"],[1571016420,"0.021918","0.02191","0.021918","0.021909","0.309","0.006771231"],[1571016480,"0.02191","0.021915","0.021916","0.02191","0.756","0.016566446"],[1571016540,"0.021915","0.021909","0.021915","0.021908","1.091","0.023905599"],[1571016600,"0.021911","0.021905","0.021912","0.021905","0.995","0.02180085"],[1571016660,"0.021904","0.021901","0.021904","0.021901","0.666","0.014586284"],[1571016720,"0.0219","0.021909","0.021909","0.0219","2.71","0.059353151"],[1571016780,"0.021913","0.021905","0.021913","0.021905","2.809","0.061540452"],[1571016840,"0.021903","0.021901","0.021903","0.0219","3.648","0.07989657"],[1571016900,"0.021901","0.021915","0.021915","0.021901","1.728","0.03785006"],[1571016960,"0.021913","0.02191","0.021913","0.021909","0.423","0.009268029"],[1571017020,"0.021907","0.021909","0.021915","0.021907","1.148","0.025155941"],[1571017080,"0.021911","0.021921","0.021921","0.021909","0.9","0.019724241"],[1571017140,"0.021922","0.02193","0.02193","0.021922","3.051","0.066895944"],[1571017200,"0.021928","0.021961","0.021969","0.021928","4.56","0.100104401"],[1571017260,"0.021965","0.021969","0.021974","0.021964","6.922","0.152062666"],[1571017320,"0.021972","0.021972","0.021975","0.02197","7.83","0.172047943"],[1571017380,"0.021972","0.021974","0.021976","0.02197","6.069","0.133357558"],[1571017440,"0.021972","0.021965","0.021975","0.021965","4.635","0.10183913"],[1571017500,"0.021964","0.02197","0.02197","0.021959","1.402","0.030792434"],[1571017560,"0.02197","0.021973","0.021973","0.021969","0.487","0.01070015"],[1571017620,"0.021972","0.021973","0.021978","0.021972","0.592","0.013008412"],[1571017680,"0.021974","0.021978","0.02198","0.021973","1.038","0.022811409"],[1571017740,"0.021976","0.021979","0.02198","0.021976","1.064","0.02338461"],[1571017800,"0.021979","0.021983","0.021983","0.021971","0.542","0.011912109"],[1571017860,"0.021976","0.021975","0.021985","0.021975","0.59","0.012967156"],[1571017920,"0.021975","0.021976","0.021979","0.021972","0.429","0.009427026"],[1571017980,"0.021978","0.021987","0.021987","0.021977","1.529","0.033612922"],[1571018040,"0.021987","0.021977","0.021992","0.021977","1.356","0.02981246"],[1571018100,"0.021977","0.021983","0.021983","0.021976","0.435","0.009560418"],[1571018160,"0.021977","0.021978","0.021983","0.021977","0.301","0.006615637"],[1571018220,"0.02198","0.021981","0.021984","0.021975","0.73","0.016045077"],[1571018280,"0.021982","0.021985","0.021985","0.021979","2.932","0.064452213"],[1571018340,"0.021987","0.021978","0.021991","0.021974","2.445","0.053742647"],[1571018400,"0.021987","0.021974","0.021987","0.021971","2.692","0.059160255"],[1571018460,"0.021976","0.021973","0.021976","0.021972","3.185","0.069985792"],[1571018520,"0.021971","0.021996","0.021997","0.021971","2.944","0.064698623"],[1571018580,"0.021995","0.021991","0.022002","0.021988","1.441","0.03168847"],[1571018640,"0.021991","0.021995","0.021996","0.021991","0.587","0.012909957"],[1571018700,"0.021994","0.021977","0.021995","0.021976","1.994","0.043839598"],[1571018760,"0.021978","0.021999","0.021999","0.021978","0.836","0.018383733"],[1571018820,"0.022009","0.022024","0.022024","0.022005","1.592","0.035038661"],[1571018880,"0.022016","0.022012","0.022018","0.022012","0.471","0.010369112"],[1571018940,"0.022012","0.022016","0.022019","0.022012","0.778","0.017127044"],[1571019000,"0.022018","0.022014","0.022019","0.022014","1.622","0.035712339"],[1571019060,"0.022016","0.022034","0.022034","0.022015","2.08","0.045808106"],[1571019120,"0.022032","0.022034","0.022039","0.022032","1.016","0.022387688"],[1571019180,"0.022033","0.022035","0.022036","0.022027","1.238","0.027277497"],[1571019240,"0.022035","0.021998","0.022035","0.021998","1.969","0.043345551"],[1571019300,"0.021993","0.022005","0.022005","0.021993","4.209","0.092591446"],[1571019360,"0.022003","0.022001","0.022003","0.021999","4.7","0.103406168"],[1571019420,"0.021998","0.022001","0.022004","0.021994","7.051","0.155106638"],[1571019480,"0.022002","0.021998","0.022004","0.021998","5.762","0.126761734"],[1571019540,"0.022003","0.022014","0.022014","0.022002","2.052","0.045159279"],[1571019600,"0.022014","0.02199","0.022014","0.021987","1.552","0.034134213"],[1571019660,"0.021993","0.022007","0.022012","0.021988","1.229","0.027039131"],[1571019720,"0.022","0.022","0.022","0.021997","1.679","0.036935778"],[1571019780,"0.022","0.02201","0.02201","0.02199","1.16","0.025517776"],[1571019840,"0.022012","0.022012","0.022016","0.02201","0.625","0.013758066"],[1571019900,"0.022016","0.02201","0.022016","0.022003","2.723","0.05992472"],[1571019960,"0.02201","0.022011","0.022014","0.022006","1.053","0.023177811"],[1571020020,"0.02201","0.02201","0.022015","0.02201","0.224","0.004930724"],[1571020080,"0.022003","0.021975","0.022005","0.021975","4.421","0.097199581"],[1571020140,"0.021959","0.021977","0.021977","0.021959","7.371","0.161910958"],[1571020200,"0.021973","0.02198","0.021984","0.021973","9.011","0.198047419"],[1571020260,"0.021976","0.02198","0.02198","0.02197","8.484","0.186447932"],[1571020320,"0.021982","0.021973","0.021982","0.021968","7.795","0.171275357"],[1571020380,"0.021972","0.021982","0.021982","0.021972","3.157","0.069374619"],[1571020440,"0.021982","0.021977","0.021985","0.021973","2.366","0.052007732"],[1571020500,"0.021975","0.021986","0.021986","0.021972","1.236","0.027165217"],[1571020560,"0.021989","0.02199","0.021992","0.021989","2.842","0.062498434"],[1571020620,"0.02199","0.021987","0.02199","0.021987","2.418","0.053168345"],[1571020680,"0.021987","0.021984","0.021987","0.021981","0.98","0.021544076"],[1571020740,"0.021983","0.021976","0.021983","0.021972","0.958","0.021054396"],[1571020800,"0.021978","0.021977","0.021984","0.021976","0.655","0.014396161"],[1571020860,"0.021978","0.021982","0.021982","0.021978","1.07","0.023519097"],[1571020920,"0.021984","0.021983","0.021988","0.02198","0.707","0.015542035"],[1571020980,"0.021983","0.021993","0.021993","0.021981","1.051","0.02310878"],[1571021040,"0.021991","0.021979","0.021993","0.021975","2.077","0.045668183"],[1571021100,"0.02198","0.021987","0.021998","0.02198","3.388","0.074499566"],[1571021160,"0.021991","0.021984","0.021991","0.021975","2.133","0.046891137"],[1571021220,"0.021988","0.021976","0.02199","0.021972","1.914","0.042083733"],[1571021280,"0.021977","0.021988","0.021994","0.021977","1.581","0.034761372"],[1571021340,"0.021987","0.021992","0.021992","0.021986","2.034","0.044727184"],[1571021400,"0.021992","0.021989","0.021997","0.021981","0.809","0.017791083"],[1571021460,"0.021991","0.021987","0.021991","0.021985","1.576","0.034653121"],[1571021520,"0.021987","0.02199","0.02199","0.021986","2.136","0.04696844"],[1571021580,"0.021987","0.021986","0.021987","0.021986","1.628","0.035794379"],[1571021640,"0.021992","0.021983","0.021992","0.02198","0.45","0.009893099"],[1571021700,"0.021982","0.022006","0.022011","0.021982","1.712","0.037664736"],[1571021760,"0.022005","0.022028","0.022033","0.022002","2.808","0.061824502"],[1571021820,"0.022027","0.022016","0.022028","0.022015","3.15","0.069373054"],[1571021880,"0.022009","0.022015","0.022015","0.022004","1.249","0.027490047"],[1571021940,"0.022015","0.022028","0.02203","0.022015","2.68","0.05902851"],[1571022000,"0.022028","0.022044","0.022062","0.022028","1.577","0.034763122"],[1571022060,"0.022048","0.022046","0.022049","0.022043","0.967","0.021318053"],[1571022120,"0.022037","0.02202","0.022042","0.02202","3.265","0.071934707"],[1571022180,"0.022025","0.02202","0.022026","0.02202","1.401","0.030854179"],[1571022240,"0.022021","0.022022","0.022024","0.022019","2.451","0.053974794"],[1571022300,"0.022022","0.02202","0.022022","0.02202","3.98","0.087643784"],[1571022360,"0.02202","0.022015","0.02202","0.022015","2.691","0.059244139"],[1571022420,"0.022019","0.022005","0.02202","0.022005","1.083","0.023839302"],[1571022480,"0.022015","0.022015","0.022019","0.022015","0.555","0.012219073"],[1571022540,"0.022017","0.022017","0.022021","0.022016","0.911","0.020058536"],[1571022600,"0.022018","0.022032","0.022032","0.022018","1.353","0.029805054"],[1571022660,"0.022029","0.022024","0.022036","0.022024","2.768","0.060984435"],[1571022720,"0.022024","0.022027","0.02203","0.022021","1.827","0.040243802"],[1571022780,"0.022026","0.022025","0.022026","0.022021","1.02","0.022464348"],[1571022840,"0.022025","0.02202","0.022027","0.02202","0.81","0.017839912"],[1571022900,"0.022023","0.02203","0.02203","0.022023","1.305","0.028744117"],[1571022960,"0.022028","0.022045","0.022049","0.022028","0.603","0.013290551"],[1571023020,"0.022051","0.022054","0.022054","0.022046","1.866","0.041147221"],[1571023080,"0.022049","0.022038","0.022049","0.022037","2.448","0.053964084"],[1571023140,"0.022035","0.022009","0.022041","0.022009","0.791","0.017430115"],[1571023200,"0.022009","0.022006","0.022016","0.022006","0.625","0.013756489"],[1571023260,"0.022009","0.022006","0.022009","0.022006","2.552","0.056163127"],[1571023320,"0.022016","0.022011","0.022016","0.022004","4.333","0.095358722"],[1571023380,"0.021998","0.022005","0.022014","0.021996","4.534","0.099759403"],[1571023440,"0.022008","0.022005","0.022008","0.022005","0.5","0.011003475"]]}
```

Reply:
Судя по всему, порядок на бирже OCHLV:

```
    parseOHLCV (ohlcv, market = undefined, timeframe = '1m', since = undefined, limit = undefined) {
        return [
            ohlcv[0] * 1000, // timestamp
            parseFloat (ohlcv[1]), // open
            parseFloat (ohlcv[3]), // high
            parseFloat (ohlcv[4]), // low
            parseFloat (ohlcv[2]), // close
            parseFloat (ohlcv[5]), // volume
        ];
    }
```

Reply:
Да, ты прав. Только что поправил это.